keras_cv/models/object_detection/faster_rcnn.py

```diff
@@ -346,81 +345,165 @@ def _call_rpn(self, images, training=None):
-            (
-                rois,
-                rcnn_box_targets,
-                rcnn_box_weights,
-                rcnn_cls_targets,
-                rcnn_cls_weights,
-            ) = self.roi_sampler(rois, gt_boxes, gt_classes)
-        feature_map = self.roi_pooler(feature_map, rois)
-        # [BS, H*W*K, pool_shape*C]
-        feature_map = tf.reshape(
-            feature_map, tf.concat([tf.shape(rois)[:2], [-1]], axis=0)
+        rpn_box_weights /= self.rpn_labeler.samples_per_image * global_batch * 0.25
+        rpn_cls_weights /= self.rpn_labeler.samples_per_image * global_batch
+        rois, feature_map, rpn_box_pred, rpn_cls_pred = self._call_rpn(
+            images, anchors, training=training
+        )
+        rois = tf.stop_gradient(rois)
+        rois, box_targets, box_weights, cls_targets, cls_weights = self.roi_sampler(
+            rois, gt_boxes, gt_classes
+        )
+        box_weights /= self.roi_sampler.num_sampled_rois * global_batch * 0.25
+        cls_weights /= self.roi_sampler.num_sampled_rois * global_batch
+        box_pred, cls_pred = self._call_rcnn(rois, feature_map)
+        y_true = {
+            "rpn_box": rpn_box_targets,
+            "rpn_cls": rpn_cls_targets,
+            "box": box_targets,
+            "cls": cls_targets,
+        }
+        y_pred = {
+            "rpn_box": rpn_box_pred,
+            "rpn_cls": rpn_cls_pred,
+            "box": box_pred,
+            "cls": cls_pred,
+        }
+        weights = {
+            "rpn_box": rpn_box_weights,
+            "rpn_cls": rpn_cls_weights,
+            "box": box_weights,
+            "cls": cls_weights,
+        }
+        return super().compute_loss(
+            x=images, y=y_true, y_pred=y_pred, sample_weight=weights
         )
-        # [BS, H*W*K, 4], [BS, H*W*K, num_classes + 1]
-        rcnn_box_pred, rcnn_cls_pred = self.rcnn_head(feature_map)
-        res = {}
-        if training:
-            res["rcnn_box_pred"] = rcnn_box_pred
-            res["rcnn_cls_pred"] = rcnn_cls_pred
-            res["rpn_box_pred"] = tf.concat(tf.nest.flatten(rpn_boxes), axis=1)
-            res["rpn_cls_pred"] = tf.concat(tf.nest.flatten(rpn_scores), axis=1)
-            res["rcnn_box_targets"] = rcnn_box_targets
-            res["rcnn_box_weights"] = rcnn_box_weights
-            res["rcnn_cls_targets"] = rcnn_cls_targets
-            res["rcnn_cls_weights"] = rcnn_cls_weights
-        else:
-            # now it will be on "center_yxhw" format
-            rcnn_box_pred = _decode_deltas_to_boxes(
-                anchors=rois,
-                boxes_delta=rcnn_box_pred,
-                anchor_format="yxyx",
-                box_format=self.bounding_box_format,
-                variance=[0.1, 0.1, 0.2, 0.2],
-            )
-            res["rcnn_box_pred"] = rcnn_box_pred
-            res["rcnn_cls_pred"] = rcnn_cls_pred
 
-        return res
+    def train_step(self, data):
+        images = data["images"]
```

Comment:
instead of data being a dictionary, lets make data a tuple of (x, y) like a standard Keras model.

Let's make `y` a dictionary with keras {"boxes", "classes"}.  We can migrate RetinaNet to the same format, alongside COCO metrics, and all other components (KPLs, etc)

Reply:
Good point. Done.